orchagent/nhgorch.cpp

```diff
@@ -0,0 +1,1192 @@
+                * actual group when there are enough resources.
+                */
+                if ((nhg_key.getSize() > 1) && (NextHopGroup::getCount() >= m_maxNhgCount))
+                {
+                    SWSS_LOG_WARN("Next hop group count reached its limit.");
+
+                    try
+                    {
+                        auto nhg = std::make_unique<NextHopGroup>(
+                                                    createTempNhg(nhg_key));
+                        if (nhg->sync())
+                        {
+                            m_syncdNextHopGroups.emplace(index,
+                                                    NhgEntry(std::move(nhg)));
+                        }
+                        else
+                        {
+                            SWSS_LOG_WARN("Failed to sync temporary NHG %s with %s",
+                                index.c_str(),
+                                nhg_key.to_string().c_str());
+                        }
+                    }
+                    catch (const std::exception& e)
+                    {
+                        SWSS_LOG_WARN("Got exception: %s while adding temp group %s",
+                            e.what(),
+                            nhg_key.to_string().c_str());
+                    }
+
+                    /*
+                     * We set the success to false so we keep trying to update
+                     * this group in order to sync the whole group.
+                     */
+                    success = false;
+                }
+                else
+                {
+                    auto nhg = std::make_unique<NextHopGroup>(nhg_key);
+                    success = nhg->sync();
+
+                    if (success)
+                    {
+                        m_syncdNextHopGroups.emplace(index,
+                                                    NhgEntry(std::move(nhg)));
+                    }
+                }
+            }
+            /* If the group exists, update it. */
+            else
+            {
+                const auto& nhg_ptr = nhg_it->second.nhg;
+
+                /*
+                 * A NHG update should never change the SAI ID of the NHG if it
+                 * is still referenced by some other objects, as they would not
+                 * be notified about this change.  The only exception to this
+                 * rule is for the temporary NHGs, as the referencing objects
+                 * will keep querying the NhgOrch for any SAI ID updates.
+                 */
+                if (!nhg_ptr->isTemp() &&
+                    ((nhg_key.getSize() == 1) || (nhg_ptr->getSize() == 1)) &&
+                    (nhg_it->second.ref_count > 0))
+                {
+                    SWSS_LOG_WARN("Next hop group %s update would change SAI "
+                                  "ID while referenced, so not performed",
+                                  index.c_str());
+                    success = false;
```

Comment:
I'm not sure I understood this logic. If either the existing NHG or the new NHG update has only 1 member, are you skipping the update? Will keep trying until the reference count of the existing NGH is '0'?

Reply:
Yes, you're right. We can't perform the update as that will change the SAI ID of the NHG object and the referencing objects would not know about that, so we'll skip the update until the NHG is not referenced anymore.

Reply:
Isn't that a concern for you? If such next hop updates don't happen there will be traffic loss until it meets the condition set. It may be a valid case of a route pointing to a valid next hop, in which case condition may never meet. 
In this case one solution would be to always use nexthop group even for a single nexthop.

Reply:
Yes, you are right. How about updating the NHG split design doc to specify that the routing stack must not update a NH to an ECMP or vice-versa while a route is pointing to the given NH/ECMP? The other option I guess would be to allow the update, but notify all referencing objects about this change so they could update the used SAI ID.

Reply:
Not updating NH to an ECMP may also cause traffic drop if the nhop members are different. In our design there are no backward references to routes for a given next hop, other than walking through the entire routing table. Tom, since you contributed to design doc, how do you plan to address this scenario. Creating a next hop group even for a single NHOP should work, there could be redundant entires between nhOrch and routeOrch, which is fine as they will be independent of each other. In case of next hop group exhaustion no nhop will be created.

Reply:
I've changed the behaviour here to create a NHG regardless of whether there's 1 or more next hops in the group - this allows easy change between the 2 and makes the code simpler